In [7]:
import numpy as np
import re

In [8]:
def read_input(filename):
    '''read the input and split by line, get rid of empty line, and return as a list of string'''
    temp_file = open(filename)
    return temp_file.read().split('\n')[:-1]

In [9]:
input_3 = read_input('input_p3.txt')

In [12]:
def find_range(input_3):
    '''find the range of area co-ordinate from 0,0 to ... so that we can find a grid large enough for plotting all co-ordinates'''
    most_down_list = []
    most_right_list = []
    right_start_list = list(map(int,re.findall(r'(?<=@ ).*?(?=,)',str(input_3))))
    down_start_list = list(map(int,re.findall(r'(?<=,).*?(?=:)',''.join(input_3))))
    right_len_list = list(map(int,re.findall(r'(?<=: ).*?(?=x)',str(input_3))))
    down_len_list = list(map(int,re.findall(r"(?<=x).*?(?=')", str(input_3))))
    for i in range(len(down_start_list)):
        most_down_list.append(down_start_list[i] + down_len_list[i])
        most_right_list.append(right_start_list[i] + right_len_list[i])
    return max(most_down_list), max(most_right_list)

In [13]:
find_range(input_3)

(1000, 999)

In [14]:
def find_overlap(input_x):
    '''find the overlapping area by numpy matrix grid'''
    grid = np.zeros([1000,1000])
    down_start = list(map(int,re.findall(r'(?<=@ ).*?(?=,)',str(input_x))))
    right_start = list(map(int,re.findall(r'(?<=,).*?(?=:)',''.join(input_x))))
    down_len = list(map(int,re.findall(r'(?<=: ).*?(?=x)',str(input_x))))
    right_len = list(map(int,re.findall(r"(?<=x).*?(?=')", str(input_x))))
    for i in range(len(down_start)):
        grid[down_start[i]:down_start[i]+down_len[i],right_start[i]:right_start[i]+right_len[i]] += 1
    return sum(grid.ravel() >1)

In [19]:
#test case
input_4 = ['#1 @ 1,3: 4x4',
           '#2 @ 3,1: 4x4',
           '#3 @ 4,4: 3x3',
          '#4 @ 204,204: 3x3']
find_overlap(input_4)

8

In [16]:
find_overlap(input_3)

111485

The answer to the first question is 111485 sq. inches overlapping.

In [17]:
def find_not_overlap(input_x):
    """use the grid system to find non-overlapping area"""
    grid = np.zeros([1000,1000])
    down_start = list(map(int,re.findall(r'(?<=@ ).*?(?=,)',str(input_x))))
    right_start = list(map(int,re.findall(r'(?<=,).*?(?=:)',''.join(input_x))))
    down_len = list(map(int,re.findall(r'(?<=: ).*?(?=x)',str(input_x))))
    right_len = list(map(int,re.findall(r"(?<=x).*?(?=')", str(input_x))))
    area_list = []
    for i in range(len(down_start)):
        grid[down_start[i]:down_start[i]+down_len[i],right_start[i]:right_start[i]+right_len[i]] += 1
        area_list.append(down_len[i]*right_len[i])
    for i in range(len(down_start)):
        if sum(grid[down_start[i]:down_start[i]+down_len[i],right_start[i]:right_start[i]+right_len[i]].ravel()) == area_list[i]:
            return i+1

In [18]:
find_not_overlap(input_3)

113

The answer to the second part is the rectangle 113 does not have any overlap with any other rectangles.